UPGRADE_GUIDE.md

```diff
@@ -2,7 +2,203 @@
+      code)
+    - `EasyPost.Exceptions.UnexpectedHttpError` - thrown when an unknown HTTP error occurs (unexpected 5xx status code)
+    - `EasyPost.Exceptions.UnknownApiError` - thrown when an unknown API error occurs (unexpected 4xx status code)
+    - `EasyPost.Exceptions.VcrError` - thrown when an error occurs with the VCR (used for testing only)
+- Generic exceptions
+    - `EasyPost.Exceptions.FilteringError` - thrown when an error occurs during filtering (e.g. calculating lowest rate)
+    - `EasyPost.Exceptions.InvalidObjectError` - thrown when an invalid object is being used
+    - `EasyPost.Exceptions.JsonError`
+        - `EasyPost.Exceptions.JsonDeserializationError` - thrown when an error occurs during JSON deserialization
+        - `EasyPost.Exceptions.JsonSerializationError` - thrown when an error occurs during JSON serialization
+        - `EasyPost.Exceptions.JsonNoDataError` - thrown when no data is provided for JSON deserialization
+    - `EasyPost.Exceptions.MissingPropertyError` - thrown when a required property is missing (e.g. `Id` for most
+      objects)
+    - `EasyPost.Exceptions.SignatureVerificationError` - thrown when the signature for a webhook is invalid
+    - `EasyPost.Exceptions.ValidationError`
+        - `EasyPost.Exceptions.MissingParameterError` - thrown when a required parameter is missing
+        - `EasyPost.Exceptions.InvalidParameterError` - thrown when an invalid parameter is passed to a function
+
+Any exception thrown by the EasyPost library will be one of the above types.
+
+Any `EasyPost.Exceptions.ApiError` exception will have the following properties:
+
+- `int? StatusCode` - the HTTP status code returned by the API call (e.g. 404)
+- `string? Code` - the error code returned by the API (e.g. "PARAMETER.INVALID")
+- `List<EasyPost.Error>? Errors` - a list of errors returned by the API (e.g. "Invalid parameter: to_address")
+
+Any `EasyPost.Exceptions.ApiError` exception can be pretty-printed using the `PrettyPrint` string property.
+
+Users can better anticipate exception information through utilities in the `EasyPost.Exceptions.Constants` namespace.
+
+- Check what exception type will be thrown for a given HTTP status code by
+  using `EasyPost.Exceptions.Constants.GetEasyPostExceptionType(HttpStatusCode statusCode)`
+- Check error message strings/templates in `EasyPost.Exceptions.Constants.ErrorMessages`
+
+## 4.0 Beta Client
+
+*Likelihood of Impact: **Medium***
+
+Following the [client instance redesign](#40-client-instance), beta features are now accessed via the `Beta` property of
+the `Client` instance.
+
+```csharp
+// Old
+var betaObject = await EasyPost.Beta.SomeService.SomeFunction(parameters);
+
+// New
+var betaObject = await myClient.Beta.SomeService.SomeFunction(parameters);
+```
+
+Any beta function will use the same client (and associated API key) from which the beta function was accessed (
```

Comment:
Thought: I'd add info here detailing that this client can/should only be used for beta functions and that a separate client should be made in the general namespace for everything else.

Reply:
Resolved, although this isn't really an issue. The beta client only has access to beta features (you can't call a non-beta feature off of a beta client). If you somehow did pass a beta client into a non-beta API call, it wouldn't matter (the beta client isn't actually what dictates the beta endpoint, that's the functions themselves. The beta client really is just a clone of the normal client.)

Reply:
Ah yes, thanks